contrib/registry/file/file_watcher.go

```diff
@@ -17,19 +18,26 @@ type Watcher struct {
 	prefix    string            // Watched prefix key, not file name prefix.
 	discovery gsvc.Discovery    // Service discovery.
 	ch        chan gsvc.Service // Changes that caused by inotify.
+	closed    atomic.Value      // Whether the channel has been closed
 }
 
 // Proceed proceeds watch in blocking way.
 // It returns all complete services that watched by `key` if any change.
 func (w *Watcher) Proceed() (services []gsvc.Service, err error) {
-	<-w.ch
-	return w.discovery.Search(context.Background(), gsvc.SearchInput{
-		Prefix: w.prefix,
-	})
+	if w.closed.Load() == nil {
+		<-w.ch
+		return w.discovery.Search(context.Background(), gsvc.SearchInput{
+			Prefix: w.prefix,
+		})
+	}
+	return nil, context.Canceled
```

Comment:
It might not appropriate using `context.Canceled` error here.
```go
if w.closed.Load() != nil {
    return nil, gerror.New("discovery service was closed")
}
<-w.ch
return w.discovery.Search(context.Background(), gsvc.SearchInput{
    Prefix: w.prefix,
})
```

Reply:
I looked at the context and this error is not used either:

https://github.com/gogf/gf/blob/9143b26182f49c6cb542e564721f89b3008ed92c/contrib/rpc/grpcx/internal/resolver/resolver_resolver.go#L58-L66

This will continue to run regardless of any abnormalities:

https://github.com/gogf/gf/blob/9143b26182f49c6cb542e564721f89b3008ed92c/contrib/rpc/grpcx/internal/resolver/resolver_resolver.go#L51-L67

So, I don't know what type of err to output.